# 数据收集

本报告的数据来源于3个不同的维度
* 项目提供的twitter_archive_enhanced.csv，其中包含了待分析的基本数据
* 项目提供的tweet_json.txt，其中包含了每条推特信息的转发数和点赞数等额外信息
* 项目同时提供了一个图像预测文件image-predictions.tsv的下载链接（https://raw.githubusercontent.com/udacity/new-dand-advanced-china/master/%E6%95%B0%E6%8D%AE%E6%B8%85%E6%B4%97/WeRateDogs%E9%A1%B9%E7%9B%AE/image-predictions.tsv）

其中twitter_archive_enhanced.csv和tweet_json.txt直接拷贝到了项目目录下
而image-predictions.tsv会使用RequestAPI的方式从网络中获取


In [27]:
# 先通过RequestAPI下载对应的文件，并且将原始文件存储到data_gathering目录中

import requests
import os
import pandas as pd
import json
import time

url = 'https://raw.githubusercontent.com/udacity/new-dand-advanced-china/master/%E6%95%B0%E6%8D%AE%E6%B8%85%E6%B4%97/WeRateDogs%E9%A1%B9%E7%9B%AE/image-predictions.tsv'
r = requests.get(url)

folder_name = 'data_gathering'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)

with open(folder_name + '/image-predictions.tsv', 'wb') as f:
    f.write(r.content)

images = pd.read_csv('data_gathering/image-predictions.tsv', sep = '\t', encoding = 'utf-8')
images
images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [28]:
# 读取对应的twitter_archive_enhanced.csv文件
archive = pd.read_csv('twitter-archive-enhanced.csv', encoding = 'utf-8')


In [29]:
#读取对应的tweet_json.txt文件，并且将数据按照需求，进行获取
#我们在twitter_archive文件中有的tweet才去获取它的扩展信息
with open('tweet_json.txt') as f:
    content = f.readlines()
    
json_list = [json.loads(line) for line in open('tweet_json.txt')]
json_tweet = [{'tweet_id': int(tweet_obj['id']),
        'date_time': pd.to_datetime(tweet_obj['created_at']),
        'favorites': tweet_obj['favorite_count'],
        'retweets': tweet_obj['retweet_count'],
        'user_followers': tweet_obj['user']['followers_count'],
        'user_favourites': tweet_obj['user']['favourites_count']} for tweet_obj in json_list]


json_tweet[9]

{'tweet_id': 890240255349198849,
 'date_time': Timestamp('2017-07-26 15:59:51'),
 'favorites': 32467,
 'retweets': 7684,
 'user_followers': 3768792,
 'user_favourites': 120162}

In [30]:
# 将处理过的字段的json文件转换成csv文件，便于分析
json_tweet = pd.DataFrame(json_tweet, columns = ['tweet_id', 'date_time', 'favorites', 'retweets',
                                               'user_followers', 'user_favourites'])

json_tweet.to_csv('new_tweet_json.txt', encoding = 'utf-8', index=False)
json_tweets_csv = pd.read_csv('new_tweet_json.txt', encoding = 'utf-8')


In [19]:
json_tweets_csv.user_followers.sort_values()

json_tweets_csv.favorites.sort_values()

json_tweets_csv.retweets.sort_values()

288         0
1291        2
271         3
339         3
112         3
29          4
1076        6
54          7
424        10
63         10
2294       14
2335       15
183        17
1519       19
176        20
209        23
2315       23
2185       25
1233       26
185        28
406        30
100        31
608        32
2220       34
2256       34
2255       37
963        37
282        37
2333       37
881        38
        ...  
935     23870
307     23870
526     23959
447     23959
886     24183
1073    24183
651     24370
1621    24370
152     24997
114     27502
621     27586
1762    30797
456     31140
300     31140
1826    31810
166     32589
130     32705
133     32705
866     33230
162     33231
534     40437
446     42045
443     42045
65      45655
410     47958
814     52101
1075    52101
531     56373
257     56373
1035    79116
Name: retweets, Length: 2352, dtype: int64

# 评估数据

在收集到数据之后，我们会通过编程和肉眼分析的方式，来初步定为数据的整洁度和数据质量问题

In [11]:
archive.info()
archive
archive['name'].value_counts()

archive['rating_denominator'].value_counts()
archive['rating_numerator'].value_counts()
archive['score_ratio'] = archive['rating_numerator']/archive['rating_denominator']

#archive.sample(5)

archive[archive['in_reply_to_user_id'].notnull()].sample(4)


archive['rating_numerator'].value_counts()
archive['rating_denominator'].value_counts()

archive.describe()


archive[archive['rating_numerator']==1776]

archive[archive.tweet_id.duplicated()]

archive.rating_numerator.sort_values()

archive

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [12]:
# 代码方式看看image数据集
images
images.info()
images['jpg_url'].value_counts()
images[images['jpg_url'] == 'https://pbs.twimg.com/media/CiibOMzUYAA9Mxz.jpg']

images[images['jpg_url'].duplicated()]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


tweet_id                                            jpg_url  \
1297  752309394570878976  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
1315  754874841593970688    https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg   
1333  757729163776290825    https://pbs.twimg.com/media/CWyD2HGUYAQ1Xa7.jpg   
1345  759159934323924993    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
1349  759566828574212096    https://pbs.twimg.com/media/CkNjahBXAAQ2kWo.jpg   
1364  761371037149827077  https://pbs.twimg.com/tweet_video_thumb/CeBym7...   
1368  761750502866649088    https://pbs.twimg.com/media/CYLDikFWEAAIy1y.jpg   
1387  766078092750233600    https://pbs.twimg.com/media/ChK1tdBWwAQ1flD.jpg   
1407  770093767776997377    https://pbs.twimg.com/media/CkjMx99UoAM2B1a.jpg   
1417  771171053431250945    https://pbs.twimg.com/media/CVgdFjNWEAAxmbq.jpg   
1427  772615324260794368    https://pbs.twimg.com/media/Cp6db4-XYAAMmqL.jpg   
1446  775898661951791106    https://pbs.twimg.com/media/CiyHLocU4AI2pJu.jpg   
1453  776819012571455488    https://pbs.twimg.com/media/CW88XN4WsAAlo8r.jpg   
1456  777641927919427584    https://pbs.twimg.com/media/CmoPdmHW8AAi8BI.jpg   
1463  778396591732486144    https://pbs.twimg.com/media/CcG07BYW0AErrC9.jpg   
1476  780496263422808064    https://pbs.twimg.com/media/Ck2d7tJWUAEPTL3.jpg   
1487  782021823840026624    https://pbs.twimg.com/media/CdHwZd0VIAA4792.jpg   
1495  783347506784731136    https://pbs.twimg.com/media/CVuQ2LeUsAAIe3s.jpg   
1510  786036967502913536    https://pbs.twimg.com/media/CtKHLuCWYAA2TTs.jpg   
1522  788070120937619456    https://pbs.twimg.com/media/Co-hmcYXYAASkiG.jpg   
1538  790723298204217344    https://pbs.twimg.com/media/CvaYgDOWgAEfjls.jpg   
1541  791026214425268224    https://pbs.twimg.com/media/CpmyNumW8AAAJGj.jpg   
1564  793614319594401792    https://pbs.twimg.com/media/CvyVxQRWEAAdSZS.jpg   
1569  794355576146903043    https://pbs.twimg.com/media/CvJCabcWgAIoUxW.jpg   
1571  794983741416415232    https://pbs.twimg.com/media/CvT6IV6WEAQhhV5.jpg   
1579  796177847564038144    https://pbs.twimg.com/media/Cwx99rpW8AMk_Ie.jpg   
1588  798340744599797760    https://pbs.twimg.com/media/CrXhIqBW8AA6Bse.jpg   
1589  798628517273620480    https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg   
1590  798644042770751489    https://pbs.twimg.com/media/CU3mITUWIAAfyQS.jpg   
1591  798665375516884993    https://pbs.twimg.com/media/CVMOlMiWwAA4Yxl.jpg   
...                  ...                                                ...   
1619  802624713319034886    https://pbs.twimg.com/media/CsrjryzWgAAZY00.jpg   
1624  803692223237865472    https://pbs.twimg.com/media/CZhn-QAWwAASQan.jpg   
1627  804413760345620481    https://pbs.twimg.com/media/CuRDF-XWcAIZSer.jpg   
1634  805958939288408065    https://pbs.twimg.com/media/CtzKC7zXEAALfSo.jpg   
1636  806242860592926720    https://pbs.twimg.com/media/Ct72q9jWcAAhlnw.jpg   
1640  807059379405148160    https://pbs.twimg.com/media/Ct2qO5PXEAE6eB0.jpg   
1645  808134635716833280    https://pbs.twimg.com/media/Cx5R8wPVEAALa9r.jpg   
1652  809808892968534016    https://pbs.twimg.com/media/CwS4aqZXUAAe3IO.jpg   
1683  813944609378369540    https://pbs.twimg.com/media/Cveg1-NXgAASaaT.jpg   
1693  816014286006976512    https://pbs.twimg.com/media/CiibOMzUYAA9Mxz.jpg   
1699  816829038950027264    https://pbs.twimg.com/media/CvoBPWRWgAA4het.jpg   
1703  817181837579653120  https://pbs.twimg.com/ext_tw_video_thumb/81596...   
1712  818588835076603904    https://pbs.twimg.com/media/Crwxb5yWgAAX5P_.jpg   
1717  819015331746349057    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1718  819015337530290176    https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg   
1727  820446719150292993    https://pbs.twimg.com/media/CxqsX-8XUAAEvjD.jpg   
1736  821813639212650496    https://pbs.twimg.com/media/CtVAvX-WIAAcGTf.jpg   
1742  822647212903690241    https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg   
1746  823269594223824897    https://pbs.twimg.com/media/C2kzTGxWEA

In [13]:
json_tweets_csv

json_tweets_csv.info()

json_tweets_csv[json_tweets_csv['tweet_id'].duplicated()]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2352 entries, 0 to 2351
Data columns (total 6 columns):
tweet_id           2352 non-null int64
date_time          2352 non-null object
favorites          2352 non-null int64
retweets           2352 non-null int64
user_followers     2352 non-null int64
user_favourites    2352 non-null int64
dtypes: int64(5), object(1)
memory usage: 110.3+ KB


Empty DataFrame
Columns: [tweet_id, date_time, favorites, retweets, user_followers, user_favourites]
Index: []

# 数据质量

数据质量从下面几个维度考虑：
* 完整性
* 有效性
* 准确性
* 一致性

### archive 表格

* 狗的名字存在一些错误的单词（such,quite,not,very,just,my,his,one,a,an）
* 需要过滤掉转发的记录，只保存原始评分的tweet记录
* 字段属性不正确（retweeted_status_id，retweeted_status_user_id）
* timestamp的数据类型应该为datetime，而不是object
* 对于“空字段”的表示，可以统一。（NaN，None）


### image 表格

* 被分析的图片的url有重复
* 数据缺失，archive中有2356条记录，而images中只有2075条记录


### tweet_json 表格

* 数据只有2352条，与archive中的数据不一致



# 数据整洁度

* 狗的种类，合并成一列，将具体的类型填写其中
* 应该将3个表格的信息进行合并
* 对于图片预测的概率（p1_conf,p2_conf,p3_conf），及预测的狗的种类(p1, p2, p3)进行合并

In [6]:
all_columns = pd.Series(list(archive) + list(images) + list(json_tweets_csv))

all_columns[all_columns.duplicated()]

17    tweet_id
29    tweet_id
dtype: object

In [15]:
list(archive)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'retweeted_status_id',
 'retweeted_status_user_id',
 'retweeted_status_timestamp',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo',
 'score_ratio']

In [16]:
list(images)

['tweet_id',
 'jpg_url',
 'img_num',
 'p1',
 'p1_conf',
 'p1_dog',
 'p2',
 'p2_conf',
 'p2_dog',
 'p3',
 'p3_conf',
 'p3_dog']

In [17]:
list(json_tweets_csv)

['tweet_id',
 'date_time',
 'favorites',
 'retweets',
 'user_followers',
 'user_favourites']

# 清理数据

根据上面分析的质量和整洁度问题，进行数据的清理

In [31]:
# 先将所有的数据列进行合并

df_all = pd.merge(archive, images, how = 'left', on = ['tweet_id'] )
df_all = pd.merge(df_all, json_tweets_csv, how = 'left', on = ['tweet_id'])

df_all.to_csv('df_all.csv', encoding = 'utf-8')
df_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 33 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
jpg_url                       20

In [32]:
# retweeted_status_id不为空，说明这条记录是被转发的记录
# 将转发的内容删除掉
df_all = df_all[pd.isnull(df_all.retweeted_status_id)]

# 对于没有提供图片分析数据的tweet记录删除掉
df_all = df_all.dropna(subset = ['jpg_url'])
len(df_all)

1994

In [33]:
# 将狗的状态进行合并
print(df_all.doggo.value_counts())
print(df_all.floofer.value_counts())
print(df_all.pupper.value_counts())
print(df_all.puppo.value_counts())

None     1920
doggo      74
Name: doggo, dtype: int64
None       1986
floofer       8
Name: floofer, dtype: int64
None      1782
pupper     212
Name: pupper, dtype: int64
None     1971
puppo      23
Name: puppo, dtype: int64


In [34]:
# 将狗的类型进行合并
melt_column = ['doggo', 'floofer', 'pupper', 'puppo']
stay_column = [x for x in df_all.columns.tolist() if x not in melt_column]

df_all = pd.melt(df_all, id_vars = stay_column, value_vars = melt_column, 
                         var_name = 'stages', value_name = 'dog_stage')


df_all = df_all.drop('stages', 1)
df_all = df_all.sort_values('dog_stage').drop_duplicates('tweet_id', keep = 'last')

print(len(df_all))

1994


In [35]:
# 计算狗狗的评分比率，得分/满分
df_all['score_ratio'] = df_all['rating_numerator']/df_all['rating_denominator']

In [36]:
df_all = df_all.drop(['retweeted_status_id','retweeted_status_user_id', 'retweeted_status_timestamp'], 1)

df_all = df_all.drop(['in_reply_to_status_id','in_reply_to_user_id'], 1)
df_all = df_all.drop(['p1','p1_conf','p2','p2_conf','p3','p3_conf','p2_dog','p3_dog','p1_dog'], 1)


df_all.to_csv('twitter_archive_master.csv', index=False, encoding = 'utf-8')

df_master = pd.read_csv('twitter_archive_master.csv')
df_master.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1994 entries, 0 to 1993
Data columns (total 17 columns):
tweet_id              1994 non-null int64
timestamp             1994 non-null object
source                1994 non-null object
text                  1994 non-null object
expanded_urls         1994 non-null object
rating_numerator      1994 non-null int64
rating_denominator    1994 non-null int64
name                  1994 non-null object
jpg_url               1994 non-null object
img_num               1994 non-null float64
date_time             1994 non-null object
favorites             1994 non-null float64
retweets              1994 non-null float64
user_followers        1994 non-null float64
user_favourites       1994 non-null float64
dog_stage             1994 non-null object
score_ratio           1994 non-null float64
dtypes: float64(6), int64(3), object(8)
memory usage: 264.9+ KB


In [37]:
# 因为想要分析转发和内容质量之间的关系，所以额外整理了一个转发数据相关的文件
# 并且按照每个月进行数据的统计

df_all['timestamp']=pd.to_datetime(df_all['timestamp'])
df_all['date_month']=pd.to_datetime(df_all['timestamp']).values.astype('datetime64[M]')

month_group = df_all.groupby('date_month')
plot_data_df = pd.DataFrame([], columns=['month','tweet_count', 'retweet_count', 'favourites_count'])

for name,group in month_group:
    nest_dict = pd.DataFrame([name,len(group.tweet_id), sum(group.retweets), sum(group.favorites)]).T
    nest_dict.columns = plot_data_df.columns
    plot_data_df = pd.concat([plot_data_df, nest_dict], ignore_index=True)
    
    
    
plot_data_df.to_csv('twitter_month_data.csv', index=False, encoding = 'utf-8')
df_master = pd.read_csv('twitter_month_data.csv')
df_master.info()    

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 4 columns):
month               22 non-null object
tweet_count         22 non-null int64
retweet_count       22 non-null float64
favourites_count    22 non-null float64
dtypes: float64(2), int64(1), object(1)
memory usage: 784.0+ bytes
